### ISSUE TITLE
Use Remote WSL over Remote SSH?

### ISSUE THREAD

**Issue by mattbell87** [microsoft/vscode#110130]
Hi,

I use VSCode remote to log in to my Windows 10 machine (which sits behind the TV). However I can't see a way to log in to my WSL2 container from there. I'd like to do some Docker container based development which requires me to use WSL2 on Win10Home.

Are we able to enable Remote WSL on a Remote SSH host? At the moment when I open the Remote Explorer I can only see local connections.

Is there a workaround?

TLDR; I would like to do this: `[Laptop VSCode] -> Remote SSH -> [Desktop VSCode Server] -> Remote WSL -> [Ubuntu]`

:)

**Comment by roblourens** [comment]
If you can set it up correctly, then you can ssh directly into your WSL host. I think that is the best option for you.

**Comment by mattbell87** [comment]
That is not as easy as it sounds for a few reasons:

1. SSH on WSL doesnt work out of the box, you need to install it, change the port and enable password authentication
2. You need to start the SSH server manualy by connecting to WSL (to start the container) and running `sudo service ssh start`
3. (And probably the biggest) There are still firewall issues with connecting to WSL2 containers externally https://github.com/microsoft/WSL/issues/4150

**Comment by mattbell87** [comment]
For anyone who finds this issue, here's how you set it up:

Connect to your Host
```sh
ssh username@windowshost    # Connect to the machine running WSL
wsl   # Start the WSL container
```

Then you'll have to do this once only (this example is for Ubuntu, adjust for your distro accordingly):
```sh
sudo apt install ssh      # Install SSHD
sudo nano /etc/ssh/sshd_config      # Open the SSHD config and uncomment/change #Port 22 to Port 2222
```

Then every time you'll have to start the SSH server on WSL
```sh
sudo service ssh start
```

Now in VSCode connect to your WSL SSH server on port 2222. 
```sh
ssh wslusername@windowshost -p 2222
```

If it doesn't work try:

- Reading this issue and trying the workarounds (if its still an open issue): https://github.com/microsoft/WSL/issues/4150
- Opening your firewall port 2222
- Setting PasswordAuthentication to yes (under wsl go back to `sudo nano /etc/ssh/sshd_config`)
- Restarting the WSL ssh server `sudo service ssh restart`

**Comment by mattbell87** [comment]
### An easier and better workaround

This will switch your Windows 10 SSH to use WSL by default.

SSH to your Windows host (SSH Server must be installed in Windows Features)
```pwsh
ssh user@windowshost
```

Start Powershell
```pwsh
powershell
```

Run this command to switch SSH from CMD to WSL
```pwsh
New-ItemProperty -Path "HKLM:\SOFTWARE\OpenSSH" -Name DefaultShell -Value "C:\WINDOWS\System32\bash.exe" -PropertyType String -Force
```

At this point you can connect to your Win10 host with Remote SSH in VSCode with your Windows username and pw. But you'll actually start and connect to WSL2.

### Credit

The workaround was inspired by:
https://www.hanselman.com/blog/the-easy-way-how-to-ssh-into-bash-and-wsl2-on-windows-10-from-an-external-machine
